presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/StateCompiler.java

```diff
@@ -669,27 +681,53 @@ else if (method.getName().startsWith("set")) {
         private final String getterName;
         private final Class<?> type;
         private final Object initialValue;
+        private final Optional<Type> sqlType;
 
-        private StateField(String name, Class<?> type, Object initialValue)
-        {
-            this(name, type, initialValue, "get" + name);
-        }
-
-        private StateField(String name, Class<?> type, Object initialValue, String getterName)
+        private StateField(String name, Class<?> type, Object initialValue, String getterName, Optional<Type> optionalSqlType)
         {
             this.name = requireNonNull(name, "name is null");
             checkArgument(!name.isEmpty(), "name is empty");
             this.type = requireNonNull(type, "type is null");
             this.getterName = requireNonNull(getterName, "getterName is null");
             this.initialValue = initialValue;
+            checkArgument(optionalSqlType != null, "sqlType is null");
+            if (optionalSqlType.isPresent()) {
+                checkArgument(optionalSqlType.equals(sqlTypeFromStackType(type)) || optionalSqlType.get().getJavaType() == type, "Stack type (%s) and provided sql type (%s) are incompatible", type.getName(), optionalSqlType.get().getDisplayName());
```

Comment:
Previously, I comment that `|| sqlType.getJavaType() == type` is redundant. And you said that it's actually needed for `byte`. I don't think it helps. `.getJavaType()` never returns `byte`. As a result, it can't be equal to `type` when `type` is `byte`.